object_oriented_peripheral_interphace/temp/man/man3/Master.3
.TH "Master" 3 "Mon Aug 12 2019" "OOPI" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Master \- A monolithic class to encapsulate and abstract the slave's communication with the master\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Master\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMaster\fP (const int SensorIDNumber, const char SensorName[], const char InstructionSet[][\fBSLAVE_COMMMAND_STRING_LENGTH\fP], const int NumberOfInstructions, const \fBsInstruct\fP MasterInstructionSet[], const int intParams[], const float floatParams[])"
.br
.ti -1c
.RI "\fBMaster\fP (void)"
.br
.ti -1c
.RI "\fB~Master\fP (void)"
.br
.ti -1c
.RI "\fBMaster\fP (volatile const \fBMaster\fP &)"
.br
.ti -1c
.RI "volatile \fBMaster\fP & \fBoperator=\fP (const \fBMaster\fP &rhs) volatile"
.br
.ti -1c
.RI "volatile \fBMaster\fP & \fBoperator=\fP (volatile const \fBMaster\fP &rhs) volatile"
.br
.ti -1c
.RI "void \fBSETUP\fP (const int SensorIDNumber, volatile char SensorName[], volatile char InstructionSet[][\fBSLAVE_COMMMAND_STRING_LENGTH\fP], const int NumberOfInstructions, volatile \fBsInstruct\fP MasterInstructionSet[], volatile int intParams[], volatile float floatParams[]) volatile"
.br
.RI "Set up of the commuication mechanism\&. "
.ti -1c
.RI "void \fBSPISetup\fP (void) volatile"
.br
.RI "Sets up SPI and attaches interrupt\&. "
.ti -1c
.RI "bool \fBHandshake\fP (void) volatile"
.br
.RI "Manages the handshake component of any transaction\&. "
.ti -1c
.RI "\fBmCmd\fP \fBloadRequest\fP (void) volatile"
.br
.RI "Retrieve the request sent by the master\&. "
.ti -1c
.RI "\fBmInstruct\fP \fBgetCurrentInstruction\fP (void) volatile"
.br
.RI "Returns the mInstruct component of the most recently loaded request sent by the master\&. "
.ti -1c
.RI "int \fBgetCurrentInstructionIntParameter\fP (void) volatile"
.br
.RI "Returns the integer parameter of the most recently loaded request sent by the master\&. "
.ti -1c
.RI "float \fBgetCurrentInstructionFloatParameter\fP (void) volatile"
.br
.RI "Returns the floating point parameter of the most recently loaded request sent by the master\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsCmd\fP Reply) volatile"
.br
.RI "Send an \fBsCmd\fP object in reply to the request recieved from \fBMaster\fP\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsInstruct\fP Instruction) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsInstruct\fP Instruction, volatile char InstructionString[\fBSLAVE_COMMMAND_STRING_LENGTH\fP]) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsInstruct\fP Instruction, const int iParam) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsInstruct\fP Instruction, const float fParam) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (const \fBsInstruct\fP Instruction, const int iParam, const int fParam) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (\fBsInstruct\fP Instruction, int iParam, volatile char InstructionString[\fBSLAVE_COMMMAND_STRING_LENGTH\fP]) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (\fBsInstruct\fP Instruction, float fParam, volatile char InstructionString[\fBSLAVE_COMMMAND_STRING_LENGTH\fP]) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendReply\fP (\fBsInstruct\fP Instruction, int iParam, float fParam, volatile char InstructionString[\fBSLAVE_COMMMAND_STRING_LENGTH\fP]) volatile"
.br
.RI "Send a reply to the request recieved from master\&. "
.ti -1c
.RI "void \fBsendData\fP (void) volatile"
.br
.RI "Sends the local \fBData\fP object to the master in reply to appropriate request\&. "
.ti -1c
.RI "void \fBsendIdentity\fP (void) volatile"
.br
.RI "Sends slave \fBIdentity\fP object to the master in reply to appropriate request\&. "
.ti -1c
.RI "bool \fBPushMeasurementVector\fP (const \fBMeasurementVectors\fP VectorNumber, const float Measurement) volatile"
.br
.RI "Pushes a data point onto one of the data vectors\&. "
.ti -1c
.RI "bool \fBPopMeasurementVector\fP (\fBMeasurementVectors\fP VectorNumber) volatile"
.br
.RI "Pops a data point from the tail end of a designated vector\&. "
.ti -1c
.RI "void \fBClearMeasurementVector\fP (\fBMeasurementVectors\fP VectorNumber) volatile"
.br
.RI "Clears all data points from a particular vector\&. "
.ti -1c
.RI "void \fBsetMeasurementVectorHeading\fP (\fBMeasurementVectors\fP VectorNumber, volatile char Heading[\fBROW_HEADING_LENGTH\fP]) volatile"
.br
.RI "Sets the string heading assigned to a paritcular data vector\&. "
.ti -1c
.RI "void \fBsetMeasurementVectorUnits\fP (\fBMeasurementVectors\fP VectorNumber, volatile char Units[\fBROW_UNIT_LENGTH\fP]) volatile"
.br
.RI "Sets the string Units assigned to a paritcular data vector\&. "
.ti -1c
.RI "bool \fBisThereData\fP (void) volatile"
.br
.RI "Checks to see whether any data has been pushed to any of the data vectors locally\&. "
.ti -1c
.RI "void \fBsendTotalNumOfInstructions\fP (void) volatile"
.br
.RI "Sends a reply to \fBMaster\fP specifiying the total number of instructions in a measurement procedure\&. "
.ti -1c
.RI "int \fBgetCurrentInstructionNumber\fP (void) volatile"
.br
.RI "Gets the value of the counter which tracks the current instruction number throughout the measurement procedure\&. "
.ti -1c
.RI "bool \fBsendNextUserInstruction\fP (void) volatile"
.br
.RI "Sends the next instruction in the measurement cycle to the master\&. "
.ti -1c
.RI "void \fBresendCurrentUserInstruction\fP (void) volatile"
.br
.RI "Resends the cuurent instruction in the measurement procedure to the master\&. "
.ti -1c
.RI "void \fBrestartUserInstructionCycle\fP (void) volatile"
.br
.RI "Restarts the measurement prcedure\&. "
.ti -1c
.RI "void \fBbeginMeasurement\fP (void) volatile"
.br
.RI "Updates state to indicate that \fBMaster\fP has requested the intitiation of the measurement procedure\&. "
.ti -1c
.RI "bool \fBshallIStart\fP (void) volatile"
.br
.RI "Checks whether the \fBbeginMeasurement()\fP method has been called\&. "
.in -1c
.SH "Detailed Description"
.PP 
A monolithic class to encapsulate and abstract the slave's communication with the master\&. 

A single transaction is characterised by the following flow of control: Clear SS -> Enter IRQ -> Recieve '?' from master -> send 'ACK' (0x06) to master -> recieve request \fBmCmd\fP -> send sCmd/Data/Identity as expected -> exit IRQ\&. Where \fBmCmd\fP, \fBsCmd\fP, \fBData\fP and \fBIdentity\fP are structures defined as types\&. The intial Recieve '?' -> send 'ACK' is known as the handshake\&. This class' responsibilities include intitialising and handling the SPI, attaching the interrupt and providing the IRQ, managing the handshake and encapsulating the reponse mechanism\&. Note the use of a volatile interface to allow for safe use of the interrupt\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBMaster\fP (const int SensorIDNumber, const char SensorName[], const char InstructionSet[][SLAVE_COMMMAND_STRING_LENGTH], const int NumberOfInstructions, const \fBsInstruct\fP MasterInstructionSet[], const int intParams[], const float floatParams[])"

.SS "\fBMaster\fP (void)"

.SS "~\fBMaster\fP (void)"

.SS "\fBMaster\fP (volatile const \fBMaster\fP & rhs)"

.SH "Member Function Documentation"
.PP 
.SS "void beginMeasurement (void) volatile"

.PP
Updates state to indicate that \fBMaster\fP has requested the intitiation of the measurement procedure\&. Designed to allow the IRQ to update the slave state so that the procedural code in main() can initiate the measurement procedure\&. 
.SS "void ClearMeasurementVector (\fBMeasurementVectors\fP VectorNumber) volatile"

.PP
Clears all data points from a particular vector\&. 
.PP
\fBParameters:\fP
.RS 4
\fIVectorNumber\fP is the enumerated reference to the row/vector in the data array being accessed\&. 
.RE
.PP

.SS "\fBmInstruct\fP getCurrentInstruction (void) volatile"

.PP
Returns the mInstruct component of the most recently loaded request sent by the master\&. 
.SS "float getCurrentInstructionFloatParameter (void) volatile"

.PP
Returns the floating point parameter of the most recently loaded request sent by the master\&. 
.SS "int getCurrentInstructionIntParameter (void) volatile"

.PP
Returns the integer parameter of the most recently loaded request sent by the master\&. 
.SS "int getCurrentInstructionNumber (void) volatile"

.PP
Gets the value of the counter which tracks the current instruction number throughout the measurement procedure\&. 
.SS "bool Handshake (void) volatile"

.PP
Manages the handshake component of any transaction\&. Manages the handshake between master and slave; defined by: Recieve '?' -> send 'ACK' (0x06)\&. 
.PP
\fBReturns:\fP
.RS 4
True if the handshake was successful\&. 
.RE
.PP

.SS "bool isThereData (void) volatile"

.PP
Checks to see whether any data has been pushed to any of the data vectors locally\&. 
.PP
\fBReturns:\fP
.RS 4
True if any data points exist in the data vectors locally\&. 
.RE
.PP

.SS "\fBmCmd\fP loadRequest (void) volatile"

.PP
Retrieve the request sent by the master\&. After each handshake, the master will proceed to send a request, defined by an \fBmCmd\fP object\&. This function must runafter each handshake\&. This function reassembles the \fBmCmd\fP request sent by the master, byte by byte and stores it in local memory\&. 
.PP
\fBReturns:\fP
.RS 4
the \fBmCmd\fP object sent by the master\&. 
.RE
.PP

.SS "volatile \fBMaster\fP& operator= (const \fBMaster\fP & rhs) volatile\fC [inline]\fP"

.SS "volatile \fBMaster\fP& operator= (volatile const \fBMaster\fP & rhs) volatile\fC [inline]\fP"

.SS "bool PopMeasurementVector (\fBMeasurementVectors\fP VectorNumber) volatile"

.PP
Pops a data point from the tail end of a designated vector\&. Removes the most recent data point from the vector in question by marking said data slot writeable to be the next \fBPushMeasurementVector()\fP\&. 
.PP
\fBParameters:\fP
.RS 4
\fIVectorNumber\fP is the enumerated reference to the row/vector in the data array being accessed\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
False if the vector/row in question is empty\&. 
.RE
.PP

.SS "bool PushMeasurementVector (const \fBMeasurementVectors\fP VectorNumber, const float Measurement) volatile"

.PP
Pushes a data point onto one of the data vectors\&. Adds a data point to the next available slot in a particular vector until the vector (row) in the data array is full\&. 
.PP
\fBParameters:\fP
.RS 4
\fIVectorNumber\fP is the enumerated reference to the row/vector in the data array being accessed\&. 
.br
\fIMeasurement\fP is the data point to be stored\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
False if the vector/row in question is full\&. (See DATA_ROW_LENGTH)\&. 
.RE
.PP

.SS "void resendCurrentUserInstruction (void) volatile"

.PP
Resends the cuurent instruction in the measurement procedure to the master\&. This method will NOT advance the measurement cycle\&. 
.SS "void restartUserInstructionCycle (void) volatile"

.PP
Restarts the measurement prcedure\&. Restarts the measuement procedure such that the next instruction sent will be the first instruction in the measurement cycle\&. 
.SS "void sendData (void) volatile"

.PP
Sends the local \fBData\fP object to the master in reply to appropriate request\&. Sends \fBData\fP object to the master in reponse to a request made by the master for \fBData\fP\&. NOTE: This must be used as the response to the appropriate mInstruct request\&. In paricular, (mInstruct)SendDataPlease 
.SS "void sendIdentity (void) volatile"

.PP
Sends slave \fBIdentity\fP object to the master in reply to appropriate request\&. Sends the slave \fBIdentity\fP object in response to an appropriate request made by master\&. NOTE: This must be used as the response to the appropriate mInstruct request\&. In paricular, (mInstruct)WhoAreYou 
.SS "bool sendNextUserInstruction (void) volatile"

.PP
Sends the next instruction in the measurement cycle to the master\&. Sends the next instruction in the measurement cycle array in reponse to the appropriate request from master\&. Specifically, (mInstruct)NextCommandPlease\&. This method will auto-advance the measurement cycle to the next instruction and re-define the 'current instruction'\&. 
.PP
\fBReturns:\fP
.RS 4
False if the final instruction in the measurement procedure has already been sent and the measurement procedure is now complete\&. 
.RE
.PP

.SS "void sendReply (const \fBsCmd\fP Reply) volatile"

.PP
Send an \fBsCmd\fP object in reply to the request recieved from \fBMaster\fP\&. 
.SS "void sendReply (const \fBsInstruct\fP Instruction) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. 
.SS "void sendReply (const \fBsInstruct\fP Instruction, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. Note the need for the char* to be defined locally as volatile char[]\&. Cannot pass string literals\&. 
.SS "void sendReply (const \fBsInstruct\fP Instruction, const int iParam) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. 
.SS "void sendReply (const \fBsInstruct\fP Instruction, const float fParam) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. 
.SS "void sendReply (const \fBsInstruct\fP Instruction, const int iParam, const int fParam) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. 
.SS "void sendReply (\fBsInstruct\fP Instruction, int iParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. Note the need for the char* to be defined locally as volatile char[]\&. Cannot pass string literals\&. 
.SS "void sendReply (\fBsInstruct\fP Instruction, float fParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. Note the need for the char* to be defined locally as volatile char[]\&. Cannot pass string literals\&. 
.SS "void sendReply (\fBsInstruct\fP Instruction, int iParam, float fParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile"

.PP
Send a reply to the request recieved from master\&. Assembles the \fBsCmd\fP object from the supplied parameters\&. Overloaded\&. Note the need for the char* to be defined locally as volatile char[]\&. Cannot pass string literals\&. 
.SS "void sendTotalNumOfInstructions (void) volatile"

.PP
Sends a reply to \fBMaster\fP specifiying the total number of instructions in a measurement procedure\&. 
.SS "void setMeasurementVectorHeading (\fBMeasurementVectors\fP VectorNumber, volatile char Heading[ROW_HEADING_LENGTH]) volatile"

.PP
Sets the string heading assigned to a paritcular data vector\&. NOTE: The Heading parameter must be declared locally as volatile char[]\&. Literal strings cannot be passed to this function\&. 
.PP
\fBParameters:\fP
.RS 4
\fIVectorNumber\fP is the enumerated reference to the row/vector in the data array in question\&. 
.br
\fIHeading[]\fP is the character array containing the string heading\&. 
.RE
.PP

.SS "void setMeasurementVectorUnits (\fBMeasurementVectors\fP VectorNumber, volatile char Units[ROW_UNIT_LENGTH]) volatile"

.PP
Sets the string Units assigned to a paritcular data vector\&. NOTE: The Units parameter must be declared locally as volatile char[]\&. Literal strings cannot be passed to this function\&. 
.PP
\fBParameters:\fP
.RS 4
\fIVectorNumber\fP is the enumerated reference to the row/vector in the data array in question\&. 
.br
\fIUnits[]\fP is the character array containing the string heading\&. 
.RE
.PP

.SS "void SETUP (const int SensorIDNumber, volatile char SensorName[], volatile char InstructionSet[][SLAVE_COMMMAND_STRING_LENGTH], const int NumberOfInstructions, volatile \fBsInstruct\fP MasterInstructionSet[], volatile int intParams[], volatile float floatParams[]) volatile"

.PP
Set up of the commuication mechanism\&. Intialised the auto-instantiated \fBMaster\fP object, attaches the IRQ and sets up SPI communications\&. 
.PP
\fBParameters:\fP
.RS 4
\fISensorIDNumber\fP is the identity number of th slave\&. 
.br
\fISensorName\fP is the string name of the slave\&. 
.br
\fIInstructionSet\fP is the array of strings associated with each instruction in the measuremnt cycle\&. Generally to be displayed to the user\&. 
.br

object_oriented_peripheral_interphace/Pseudo-Unit Tests/Slave_Sketch/Master.cpp

#include <Arduino.h>
#include "Master.h"
#include "SPI_InstructionSet.h"
#include <SPI.h>
#include "SPI_Anything_Slave.h"

// Constructor /////////////////////////////////////////////////////////////////
// Function that handles the creation and setup of instances

  Master::Master(){
  //   CurrentRequest = (mCmd){SitRep,0,0.0};

  //   //Initialise Identity
  //   ThisSensorID.sensorID = 0;
  //   ThisSensorID.sensorChipSelect = 0;

  //   for (int i =0; i< IDENTITY_SENSOR_NAME_LENGTH; i++){
  //     ThisSensorID.SensorName[i] = '\n';
  //   }

  // //Initialise User Instruction Set
  //   UserInstructionSet.NumOfInstructions = 0;
  //   UserInstructionSet.InstructionCounter = 0;

  //   for (int i = 0; i<MAX_USER_INSTRUCTION_NUMBER; i++){

  //     UserInstructionSet.MasterInstructionSet[i] = NAK;
  //     UserInstructionSet.iParams[i] = 0;
  //     UserInstructionSet.fParams[i] = 0.0;

  //     for (int j = 0; j<SLAVE_COMMMAND_STRING_LENGTH;j++){
  //       UserInstructionSet.InstructionSet[i][j] = '\n';
  //     }
  //   }

  // //Initialise Data
  //   //MeasurementData.NumColumns = 0;
  //   MeasurementData.NumRows = 0;

  //   for (int i = 0; i<NUMBER_OF_DATA_ROWS; i++){
  //     MeasurementData.RowHeadings[i][0] = '\n';
  //     MeasurementData.rowUnits[i][0] = '\n';
  //     MeasurementData.NumColumns[i] = 0;

  //     for (int j = 0 ; j<DATA_ROW_LENGTH; j++){
  //       MeasurementData.DataPoints[i][j] = 0.0;
  //     }
  //   }

  //Setup Slave SPI
    //SPISetup();
  }

  Master::Master(const int SensorIDNumber, const char SensorName[], const char InstructionSet[][SLAVE_COMMMAND_STRING_LENGTH], const int NumberOfInstructions, const sInstruct MasterInstructionSet[], const int intParams[], const float floatParams[]){

    beginMeasurementYet = false;

  //Initialise Identity
    ThisSensorID.sensorID = SensorIDNumber;
    ThisSensorID.sensorChipSelect = 0;

    for (int i =0; i< IDENTITY_SENSOR_NAME_LENGTH; i++){
      ThisSensorID.SensorName[i] = SensorName[i];
      if ('\n' == SensorName[i]) break;
    }

  //Initialise User Instruction Set
    UserInstructionSet.NumOfInstructions = NumberOfInstructions;
    UserInstructionSet.InstructionCounter = 0;

    for (int i = 0; i<NumberOfInstructions; i++){

      UserInstructionSet.MasterInstructionSet[i] = MasterInstructionSet[i];
      UserInstructionSet.iParams[i] = intParams[i];
      UserInstructionSet.fParams[i] = floatParams[i];

      for (int j = 0; j<SLAVE_COMMMAND_STRING_LENGTH;j++){
        UserInstructionSet.InstructionSet[i][j] = InstructionSet[i][j];
        if (InstructionSet[i][j] == '\n') break;
      }
    }

  //Initialise Data
    //MeasurementData.NumColumns = 0;
    MeasurementData.NumRows = 0;

    for (int i = 0; i<NUMBER_OF_DATA_ROWS; i++){
      MeasurementData.RowHeadings[i][0] = '\n';
      MeasurementData.rowUnits[i][0] = '\n';
      MeasurementData.NumColumns[i] = 0;

      for (int j = 0 ; j<DATA_ROW_LENGTH; j++){
        MeasurementData.DataPoints[i][j] = 0.0;
      }
    }

  //Setup Slave SPI
    //SPISetup();

}

 void Master::SETUP(const int SensorIDNumber, volatile char SensorName[], volatile char InstructionSet[][SLAVE_COMMMAND_STRING_LENGTH], const int NumberOfInstructions, volatile sInstruct MasterInstructionSet[], volatile int intParams[], volatile float floatParams[]) volatile{
    
    beginMeasurementYet = false;  

    //Initialise Identity
    ThisSensorID.sensorID = SensorIDNumber;
    ThisSensorID.sensorChipSelect = 0;

    for (int i =0; i< IDENTITY_SENSOR_NAME_LENGTH; i++){
      ThisSensorID.SensorName[i] = SensorName[i];
      if ('\n' == SensorName[i]) break;
    }

  //Initialise User Instruction Set
    UserInstructionSet.NumOfInstructions = NumberOfInstructions;
    UserInstructionSet.InstructionCounter = 0;

    for (int i = 0; i<NumberOfInstructions; i++){

      UserInstructionSet.MasterInstructionSet[i] = MasterInstructionSet[i];
      UserInstructionSet.iParams[i] = intParams[i];
      UserInstructionSet.fParams[i] = floatParams[i];

      for (int j = 0; j<SLAVE_COMMMAND_STRING_LENGTH;j++){
        UserInstructionSet.InstructionSet[i][j] = InstructionSet[i][j];
        if (InstructionSet[i][j] == '\n') break;
      }
    }

  //Initialise Data
    //MeasurementData.NumColumns = 0;
    MeasurementData.NumRows = 0;

    for (int i = 0; i<NUMBER_OF_DATA_ROWS; i++){
      MeasurementData.RowHeadings[i][0] = '\n';
      MeasurementData.rowUnits[i][0] = '\n';
      MeasurementData.NumColumns[i] = 0;

      for (int j = 0 ; j<DATA_ROW_LENGTH; j++){
        MeasurementData.DataPoints[i][j] = 0.0;
      }
    }

  //Setup SPI and interrup

    SPISetup();
    
 }


void Master::SPISetup() volatile{

  //  // have to send on master in, *slave out*
  // pinMode(MISO, OUTPUT);

  // // turn on SPI in slave mode
  // SPCR |= _BV(SPE);

  // // turn on interrupts
  // SPCR |= _BV(SPIE);

   SPI.beginTransactionSlave(SPISettings(18000000, MSBFIRST, SPI_MODE0, DATA_SIZE_8BIT));
   attachInterrupt(SPI1_NSS_PIN , SPI_IRQ, FALLING);

}

Master::~Master(void){
  //SPI.end();
}

Master::Master(volatile const Master& rhs){
  Serial.println("Copy");
  Serial.println(millis());
  this->CurrentRequest = rhs.CurrentRequest;
  this->ThisSensorID = rhs.ThisSensorID;
  this->MeasurementData = rhs.MeasurementData;
  this->UserInstructionSet = rhs.UserInstructionSet;;
}


// Public Methods //////////////////////////////////////////////////////////////
// Functions available in Wiring sketches, this library, and other libraries
  //Communications Interface
      bool Master::Handshake(void) volatile{
        // byte Ping = SPDR;
        uint8_t Ping = SPI.read();
        if (Ping == '?'){
          SPI.transfer(0x06);
          return true;
        }else{
          return false;
        }
      }

      mCmd Master::loadRequest(void) volatile{
        SPI_read(CurrentRequest);
        return CurrentRequest;
      }

      mInstruct Master::getCurrentInstruction(void) volatile{
        return CurrentRequest.Instruction;
      }

      int Master::getCurrentInstructionIntParameter(void) volatile{
        return CurrentRequest.iParam;
      }

      float Master::getCurrentInstructionFloatParameter(void) volatile{
        return CurrentRequest.fParam;
      }

      void Master::sendReply(const sCmd Reply) volatile{
        // SPI_write(Reply);
        SPI_writeAnything(Reply);
      }

      void Master::sendReply(const sInstruct Instruction) volatile{
        sCmd Reply = {Instruction,"",0,0.0};
      
        SPI_write(Reply);
      }

      void Master::sendReply(const sInstruct Instruction, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile{
        sCmd Reply = {Instruction,"",0,0.0};

        for (int i = 0; i<SLAVE_COMMMAND_STRING_LENGTH; i++){
          Reply.sParam[i] = InstructionString[i];
          if (InstructionString[i] == '\n') break;
        }
      
        SPI_write(Reply);

      }

      void Master::sendReply(const sInstruct Instruction, const int iParam) volatile{
      
        SPI_write((sCmd){Instruction,"",iParam,0.0});
      
      }

      void Master::sendReply(const sInstruct Instruction, const float fParam) volatile{

        SPI_write((sCmd){Instruction,"",0,fParam});

      }

      void Master::sendReply(const sInstruct Instruction, const int iParam, const int fParam) volatile{

        SPI_write((sCmd){Instruction,"",iParam,fParam});

      }

      void Master::sendReply( sInstruct Instruction,  int iParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile{

        sCmd Reply = {Instruction,"",iParam,0.0};

        for (int i = 0; i<SLAVE_COMMMAND_STRING_LENGTH; i++){
          Reply.sParam[i] = InstructionString[i];
          if (InstructionString[i] == '\n') break;
        }

        SPI_write(Reply);

      }
      
      void Master::sendReply( sInstruct Instruction,  float fParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile{

        sCmd Reply = {Instruction,"",0,fParam};

        for (int i = 0; i<SLAVE_COMMMAND_STRING_LENGTH; i++){
          Reply.sParam[i] = InstructionString[i];
          if (InstructionString[i] == '\n') break;
        }

        SPI_write(Reply);

      }

      void Master::sendReply( sInstruct Instruction, int iParam, float fParam, volatile char InstructionString[SLAVE_COMMMAND_STRING_LENGTH]) volatile{


      sCmd Reply = {Instruction,"",iParam,fParam};

        for (int i = 0; i<SLAVE_COMMMAND_STRING_LENGTH; i++){
          Reply.sParam[i] = InstructionString[i];
          if (InstructionString[i] == '\n') break;
        }

        SPI_write(Reply);

      }


      void Master::sendData(void) volatile{
          SPI_write(MeasurementData);
      }

      void Master::sendIdentity(void) volatile{
        SPI_write(ThisSensorID);
      }

    //Data handling interface
      bool Master::PushMeasurementVector(const MeasurementVectors VectorNumber, const float Measurement) volatile{

        switch (VectorNumber){

          case First:
            if (MeasurementData.NumColumns[First] == DATA_ROW_LENGTH) return false;
            MeasurementData.DataPoints[First][MeasurementData.NumColumns[First]] = Measurement;
            MeasurementData.NumColumns[First]++;
            break;

          case Second:
            if (MeasurementData.NumColumns[Second] == DATA_ROW_LENGTH) return false;
            MeasurementData.DataPoints[Second][MeasurementData.NumColumns[Second]] = Measurement;
            MeasurementData.NumColumns[Second]++;
            break;

          case Third:
            if (MeasurementData.NumColumns[Third] == DATA_ROW_LENGTH) return false;
            MeasurementData.DataPoints[Third][MeasurementData.NumColumns[Third]] = Measurement;
            MeasurementData.NumColumns[Third]++;
            break;

        }

        return true;

      }

      bool Master::PopMeasurementVector( MeasurementVectors VectorNumber)volatile{

        switch (VectorNumber){

          case First:
            if (MeasurementData.NumColumns[First] == 0) return false;
            MeasurementData.NumColumns[First]--;
            break;

          case Second:
            if (MeasurementData.NumColumns[Second] == 0) return false;
            MeasurementData.NumColumns[Second]--;
            break;

          case Third:
            if (MeasurementData.NumColumns[Third] == 0) return false;
            MeasurementData.NumColumns[Third]--;
            break;

        }

        return true;

      }

      void Master::ClearMeasurementVector( MeasurementVectors VectorNumber)volatile{

        switch (VectorNumber){

          case First:
            MeasurementData.NumColumns[First] = 0;
            break;

          case Second:
            MeasurementData.NumColumns[Second] = 0;
            break;

          case Third:
            MeasurementData.NumColumns[Third] = 0;
            break;

        }

      }

      void Master::setMeasurementVectorHeading( MeasurementVectors VectorNumber, volatile char Heading[ROW_HEADING_LENGTH])volatile{

        switch (VectorNumber){

          case First:
            for (int i =0; i<ROW_HEADING_LENGTH; i++){
              MeasurementData.RowHeadings[First][i] = Heading[i];
              if (Heading[i] == '\n') break;
            }
            break;

          case Second:
            for (int i =0; i<ROW_HEADING_LENGTH; i++){
              MeasurementData.RowHeadings[Second][i] = Heading[i];
              if (Heading[i] == '\n') break;
            }
            break;

          case Third:
            for (int i =0; i<ROW_HEADING_LENGTH; i++){
              MeasurementData.RowHeadings[Third][i] = Heading[i];
              if (Heading[i] == '\n') break;
            }
            break;

        }

      }

      void Master::setMeasurementVectorUnits(MeasurementVectors VectorNumber, volatile char Units[ROW_UNIT_LENGTH])volatile{

        switch (VectorNumber){

          case First:
            for (int i =0; i<ROW_UNIT_LENGTH; i++){
              MeasurementData.rowUnits[First][i] = Units[i];
              if (Units[i] == '\n') break;
            }
            break;

          case Second:
            for (int i =0; i<ROW_UNIT_LENGTH; i++){
              MeasurementData.rowUnits[Second][i] = Units[i];
              if (Units[i] == '\n') break;
            }
            break;

          case Third:
            for (int i =0; i<ROW_UNIT_LENGTH; i++){
              MeasurementData.rowUnits[Third][i] = Units[i];
              if (Units[i] == '\n') break;
            }
            break;

      }

    }

      bool Master::isThereData(void) volatile{

        if ((MeasurementData.NumColumns[First] > 0) || (MeasurementData.NumColumns[Second] > 0) ||(MeasurementData.NumColumns[Third] > 0)){
          return true;
        } else{
          return false;
        }
      }

    //User Instruction Interface
      void Master::sendTotalNumOfInstructions(void) volatile{
        sendReply((sInstruct)ReferToInt,UserInstructionSet.NumOfInstructions);
      }

      int Master::getCurrentInstructionNumber(void) volatile{
        return UserInstructionSet.InstructionCounter;
      }

      bool Master::sendNextUserInstruction(void) volatile{

        if (UserInstructionSet.InstructionCounter == UserInstructionSet.NumOfInstructions){
          return false;
        }

        int Counter = UserInstructionSet.InstructionCounter;
        sendReply(UserInstructionSet.MasterInstructionSet[Counter], UserInstructionSet.iParams[Counter], UserInstructionSet.fParams[Counter], UserInstructionSet.InstructionSet[Counter]);
      
        UserInstructionSet.InstructionCounter++;

        return true;
      }

      void Master::resendCurrentUserInstruction(void) volatile{
        int Counter = UserInstructionSet.InstructionCounter;
        sendReply(UserInstructionSet.MasterInstructionSet[Counter], UserInstructionSet.iParams[Counter], UserInstructionSet.fParams[Counter], UserInstructionSet.InstructionSet[Counter]);
       }

      void Master::restartUserInstructionCycle(void) volatile{
        UserInstructionSet.InstructionCounter = 0;
      }

      void Master::beginMeasurement(void) volatile{
        beginMeasurementYet = true;
        SensorMaster.sendReply(ACK);
      }

      bool Master::shallIStart(void) volatile{
        if (beginMeasurementYet){
          beginMeasurementYet = false; //reset
          return true;
        } else{
          return beginMeasurementYet;
        }

      }


// Private Methods /////////////////////////////////////////////////////////////
// Functions only available to other functions in this library

template <typename T> unsigned int Master::SPI_write (const T& value) volatile{
  const byte * p = (const byte*) &value;
    unsigned int i;
    for (i = 0; i < sizeof value; i++)
          SPI.transfer(*p++);
    return i;
}


template <typename T> unsigned int Master::SPI_read(T& value) volatile{
  byte * p = (byte*) &value;
    unsigned int i;
    for (i = 0; i < sizeof value; i++)
          *p++ = SPI.transfer (0);
    return i;
}


////////////////////////////////////////////////////////////////////////////////////////////////////

volatile Master SensorMaster;

void SPI_IRQ(void){
  
      if (SensorMaster.Handshake()){
        
      //Retrieve the request from Master
          mCmd Request = SensorMaster.loadRequest();

      //Call Request Handler
          RequestHandler(Request);
    }//End of transaction

//////////////////////////////////////////////////////////////////////////////////////
}